marin/rl/parquet_store.py

```diff
@@ -0,0 +1,159 @@
+"""Parquet storage utilities for :pymod:`marin.rl` rollouts.
+
+The helpers defined here keep the writer/reader logic *very* thin—primarily
+bridging between the in-memory dataclass objects and Apache Arrow structures so
+we can leverage Arrow/Parquet's excellent performance and Ray integration.
+
+Design goals
+------------
+1. Zero third-party dependencies beyond ``pyarrow`` (already required by Ray).
+2. Immutable dataclasses stay immutable; conversion happens *outside* them.
+3. Files are written as independent parts (UUID filenames) so multiple actors
+   can append concurrently without coordination.  The target directory therefore
+   represents a Parquet *dataset*.
+"""
+
+import json
+import uuid
+from collections.abc import Iterator
+
+import pyarrow as pa
+import pyarrow.dataset as ds
+import pyarrow.fs as pafs
+import pyarrow.parquet as pq
+
+from .types import Rollout, RolloutGroup, Turn
+
+# ---------------------------------------------------------------------------
+# Conversion helpers
+# ---------------------------------------------------------------------------
+
+
+def _turn_to_pyobj(turn: Turn) -> dict:
+    """Convert a :class:`Turn` into a Python mapping understood by Arrow."""
+
+    return {
+        "message": turn.message,
+        "role": turn.role,
+        "logprobs": list(turn.logprobs) if turn.logprobs is not None else None,
+        "reward": turn.reward,
+        "inference_metadata_json": json.dumps(turn.inference_metadata, separators=(",", ":")),
+    }
+
+
+def _rollout_to_pyobj(rollout: Rollout) -> dict:
+    """Convert a :class:`Rollout` into a flat Python mapping suitable for Arrow."""
+
+    return {
+        "turns": [_turn_to_pyobj(t) for t in rollout.turns],
+        "rollout_metadata_json": json.dumps(rollout.metadata, separators=(",", ":")),
+    }
+
+
+def _groups_to_table(groups: list[RolloutGroup]) -> pa.Table:
+    rows = []
+    for g in groups:
+        for r in g.rollouts:
+            row = _rollout_to_pyobj(r)
+            row["id"] = g.id
+            row["source"] = g.source
+            row["created"] = g.created
+            row["group_metadata_json"] = json.dumps(g.metadata, separators=(",", ":"))
+            rows.append(row)
+
+    return pa.Table.from_pylist(rows)
+
+
+# ---------------------------------------------------------------------------
+# Public API
+# ---------------------------------------------------------------------------
+
+
+def write_rollout_groups(
```

Comment:
Right now, this is just used in tests...where does this fit into the actual RL training code?  Why write to disk / I thought we're using Ray to ship rollouts around?

Reply:
for replay, testing, posthoc inspection, whatever